File: code/dart/modular_api/test/tracing/tracing_lifecycle_test.dart
import 'package:dartastic_opentelemetry/dartastic_opentelemetry.dart' as sdk;
import 'package:dartastic_opentelemetry/testing.dart';
import 'package:http/http.dart' as http;
import 'package:modular_api/modular_api.dart';
import 'package:test/test.dart';

/// Tests for tracing's configuration and shutdown behaviour.
///
/// **There is deliberately no tracing plugin**, and this file is where that outcome is
/// pinned. Every responsibility ADR-0005 decision 4 gave one was reassigned by a later,
/// better-evidenced decision:
///
/// - **A2** moved the sampler, span processor and exporter to the application.
/// - **Decision 4 itself** made the span host-owned, so no middleware is registered.
/// - **D8 as revised** turned provider shutdown into a callback the application supplies,
///   because the OpenTelemetry API exposes no shutdown on the provider in TypeScript or
///   Python and the provider belongs to the application anyway.
///
/// What was left was a plugin whose `setup()` did nothing and whose only act was to
/// forward one callback. That is indirection with no reader benefit, so the host reads
/// `TracingOptions` directly and invokes the callback in its existing shutdown path.
///
/// Two decisions are pinned so they cannot be undone by habit:
///
/// - **D14** — no dropped-span counter. OpenTelemetry standardises that signal as
///   `otel.sdk.processor.span.processed` with `error.type = queue_full`, and the obligation
///   sits on the SDK's batching processor.
/// - **D8 as revised** — the framework never shuts down or flushes a provider it did not
///   create. It offers the timing; the application decides what that means.
void main() {
  late TestHarness harness;

  setUpAll(() async {
    harness = await maybeInitializeOtelForTest(serviceName: 'modular_api-test');
  });

  setUp(() => harness.clear());

  ModularApi apiWith({TracingOptions? tracing}) {
    final api = ModularApi(
      basePath: '/api',
      title: 'modular_api-test',
      tracing: tracing,
    );
    api.module('cuenta', (m) {
      m.usecase(
        'ping',
        PingUseCase.fromJson,
        inputExample: PingInput(value: 'ping'),
        outputExample: PingOutput(echo: 'ping'),
      );
    });
    return api;
  }

  TracingOptions options({
    Future<void> Function()? onShutdown,
    String? instrumentationName,
  }) =>
      TracingOptions(
        tracerProvider: sdk.OTel.tracerProvider(),
        onShutdown: onShutdown,
        instrumentationName: instrumentationName ?? 'modular_api',
      );

  group('shutdown timing (D8 as revised)', () {
    test('onShutdown runs when the server closes', () async {
      // On Cloud Run this is the only window before the container dies. The framework
      // supplies the moment; the application supplies the action.
      var called = 0;

      final api = apiWith(tracing: options(onShutdown: () async => called++));
      final server = await api.serve(port: 0);
      await server.close(force: true);

      expect(called, equals(1));
    });

    test('a missing onShutdown is not an error', () async {
      final api = apiWith(tracing: options());
      final server = await api.serve(port: 0);

      await expectLater(server.close(force: true), completes);
    });

    test('an onShutdown that throws does not prevent the server closing', () async {
      // A failing flush must not turn a clean shutdown into a hang or a crash. Losing
      // telemetry is bad; failing to stop is worse.
      final api = apiWith(
        tracing: options(onShutdown: () async => throw StateError('flush failed')),
      );
      final server = await api.serve(port: 0);

      await expectLater(server.close(force: true), completes);
    });

    test('the framework does not shut down the provider itself', () async {
      // D8 as revised. Dart's API happens to expose shutdown() on the provider, and
      // TypeScript's and Python's do not — so using it here would make an observable
      // behaviour differ by language (G6). It is also a resource the application created
      // and may still be using elsewhere.
      final api = apiWith(tracing: options());
      final server = await api.serve(port: 0);
      await server.close(force: true);

      // Still usable after the server closed: nothing was torn down behind the
      // application's back.
      final span = sdk.OTel.tracerProvider()
          .getTracer('after-shutdown')
          .startSpan('still-working')
        ..end();

      expect(span.spanContext.isValid, isTrue);
    });
  });

  group('D14 — no dropped-span counter', () {
    test('tracing registers no metrics of its own', () async {
      // OpenTelemetry standardises this signal as otel.sdk.processor.span.processed with
      // error.type = queue_full, and puts the obligation on the SDK's batching processor.
      // A counter of ours would be a non-standard name beside a standard one, so this test
      // exists to stop it being reintroduced by habit.
      final api = ModularApi(
        basePath: '/api',
        title: 'modular_api-test',
        metricsEnabled: true,
        tracing: options(),
      );
      api.module('cuenta', (m) {
        m.usecase(
          'ping',
          PingUseCase.fromJson,
          inputExample: PingInput(value: 'ping'),
          outputExample: PingOutput(echo: 'ping'),
        );
      });

      final server = await api.serve(port: 0);
      try {
        final metrics = await http
            .get(Uri.parse('http://localhost:${server.port}/api/metrics'));

        expect(metrics.body, isNot(contains('spans_dropped')));
        expect(metrics.body, isNot(contains('span')));
      } finally {
        await server.close(force: true);
      }
    });

    test('tracing works with metrics disabled', () async {
      // Tracing must not require the metrics surface, which it would if it owned a counter.
      final api = apiWith(tracing: options());
      final server = await api.serve(port: 0);

      try {
        final response = await http.post(
          Uri.parse('http://localhost:${server.port}/api/cuenta/ping'),
          headers: {'content-type': 'application/json'},
          body: '{"value":"ping"}',
        );

        expect(response.statusCode, equals(200));
        expect(harness.spans.spanNames, contains('POST /api/cuenta/ping'));
      } finally {
        await server.close(force: true);
      }
    });
  });

  group('configuration', () {
    test('the instrumentation name defaults to modular_api', () {
      expect(options().instrumentationName, equals('modular_api'));
    });

    test('the instrumentation name can be overridden', () {
      expect(
        options(instrumentationName: 'socia-api').instrumentationName,
        equals('socia-api'),
      );
    });
  });
}

// ─── Fixtures ──────────────────────────────────────────────────────

class PingInput extends Input {
  PingInput({required this.value});

  factory PingInput.fromJson(Map<String, dynamic> json) =>
      PingInput(value: json['value'] as String? ?? '');

  final String value;

  @override
  Map<String, dynamic> toJson() => {'value': value};

  @override
  Map<String, dynamic> toSchema() => {
        'type': 'object',
        'properties': {
          'value': {'type': 'string'},
        },
        'required': ['value'],
      };
}

class PingOutput extends Output {
  PingOutput({required this.echo});

  final String echo;

  @override
  int get statusCode => 200;

  @override
  Map<String, dynamic> toJson() => {'echo': echo};

  @override
  Map<String, dynamic> toSchema() => {
        'type': 'object',
        'properties': {
          'echo': {'type': 'string'},
        },
        'required': ['echo'],
      };
}

class PingUseCase implements UseCase<PingInput, PingOutput> {
  PingUseCase({required this.input});

  static PingUseCase fromJson(Map<String, dynamic> json) =>
      PingUseCase(input: PingInput.fromJson(json));

  @override
  final PingInput input;

  @override
  ModularLogger? logger;

  @override
  String? validate() => null;

  @override
  Future<PingOutput> execute() async => PingOutput(echo: input.value);
}

File: code/ts/modular_api/test/tracing/tracingLifecycle.test.ts
import { InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import type { Server } from 'node:http';
import { beforeAll, describe, expect, it } from 'vitest';

import { ModularApi } from '../../src/core/modular_api';
import { TracingOptions } from '../../src/core/tracing/tracingOptions';

/**
 * TypeScript mirror of `test/tracing/tracing_lifecycle_test.dart`.
 *
 * **There is deliberately no tracing plugin** (ADR-0005 A7). Every responsibility decision 4
 * assigned to one was reassigned: A2 moved the sampler, processor and exporter to the
 * application; the span became host-owned so no middleware is registered; the shutdown
 * became a callback; and D14 removed the dropped-span counter. The host reads
 * `TracingOptions` directly.
 *
 * D8's revision is starker here than in Dart: `TracerProvider` in `@opentelemetry/api`
 * exposes only `getTracer`. The framework *could not* flush the provider even if it wanted
 * to, which is half of why the responsibility moved to the application.
 */
const exporter = new InMemorySpanExporter();
let provider: NodeTracerProvider;

beforeAll(() => {
  provider = new NodeTracerProvider({ spanProcessors: [new SimpleSpanProcessor(exporter)] });
  provider.register();
});

const options = (onShutdown?: () => Promise<void>): TracingOptions =>
  new TracingOptions({
    tracerProvider: provider,
    ...(onShutdown === undefined ? {} : { onShutdown }),
  });

async function serveAndClose(tracing?: TracingOptions): Promise<void> {
  const api = new ModularApi({ basePath: '/api', title: 'modular_api-test', ...(tracing ? { tracing } : {}) });
  const server: Server = await api.serve({ port: 0 });
  await new Promise<void>((resolve) => server.close(() => resolve()));
  // `close` fires the handler asynchronously; give the microtask queue a turn.
  await new Promise<void>((resolve) => setImmediate(resolve));
}

describe('shutdown timing (D8 as revised)', () => {
  it('onShutdown runs when the server closes', async () => {
    // On Cloud Run this is the only window before the container dies. The framework
    // supplies the moment; the application supplies the action.
    let called = 0;

    await serveAndClose(options(async () => void called++));

    expect(called).toBe(1);
  });

  it('a missing onShutdown is not an error', async () => {
    await expect(serveAndClose(options())).resolves.toBeUndefined();
  });

  it('an onShutdown that rejects does not prevent the server closing', async () => {
    // A failing flush must not turn a clean shutdown into a hang or a crash.
    await expect(
      serveAndClose(
        options(async () => {
          throw new Error('flush failed');
        }),
      ),
    ).resolves.toBeUndefined();
  });

  it('the framework does not shut down the provider itself', async () => {
    // It could not: TracerProvider in @opentelemetry/api exposes only getTracer. Asserting
    // it anyway, because the test states the intent rather than the accident.
    await serveAndClose(options());

    const span = provider.getTracer('after-shutdown').startSpan('still-working');
    span.end();

    expect(span.spanContext().traceId).toBeDefined();
  });
});

describe('configuration', () => {
  it('the instrumentation name defaults to modular_api', () => {
    expect(options().instrumentationName).toBe('modular_api');
  });

  it('the instrumentation name can be overridden', () => {
    expect(
      new TracingOptions({ tracerProvider: provider, instrumentationName: 'socia-api' })
        .instrumentationName,
    ).toBe('socia-api');
  });
});
